content/Improve your Rails App searches with Elasticsearch - Leandro Dominguez.md

```diff
@@ -0,0 +1,213 @@
+
+The response should see something like this:
+
+```json
+{
+  "name" : "192.168.0.100",
+  "cluster_name" : "elasticsearch_instance",
+  "cluster_uuid" : "LjE5yneZRWOUFnlasU6y7g",
+  "version" : {
+    "number" : "7.4.2",
+    "build_flavor" : "default",
+    "build_type" : "tar",
+    "build_hash" : "2f90bbf7b93631e52bafb59b3b049cb44ec25e96",
+    "build_date" : "2019-10-28T20:40:44.881551Z",
+    "build_snapshot" : false,
+    "lucene_version" : "8.2.0",
+    "minimum_wire_compatibility_version" : "6.8.0",
+    "minimum_index_compatibility_version" : "6.0.0-beta1"
+  },
+  "tagline" : "You Know, for Search"
+}
+```
+
+Now we have our Elastic instance up and running, open a Rails console and create the corresponding index by executing:
+
+```ruby
+Post.__elasticsearch__.create_index!
+```
+
+If you already had some posts created, you will need to reindex them by running the following:
+
+```ruby
+Post.import
+```
+
+This can take A LOT of time if you have too many records, since Elastic has to analyze and tokenize all of them.
+
+**Step 5: define the search method we will call from the controller**
+
+In the Posts’ model define this basic method that will execute a query in our Elasticsearch instance, taking a string and looking in the specified fields.
+
+```ruby
+   def self.our_first_query(query)
+     search(
+       query: {
+         multi_match: {
+           query: query,
+           fields: %w[title body^5]
+         }
+       }
+     )
+   end
+```
+
+This simple query will search through all our posts and retrieve them in order of importance. The ‘`^5`’ after `body` will make a match in this attribute to weigh 5 more times than one on `title`.
+
+**Step 6: create your searches method in the controller, its view and routes.**
+
+My controller method looks just like this
+
+```ruby
+ def search
+   return unless query.present?
+
+   @posts = Post.our_first_query(query)
+ end
+```
+
+Also, I did this simple view with a little of Bootstrap.
+
+![Search by semantic information](images/semantic_search.png)
+
+As you can see, I created a bunch of posts using [faker]([https://github.com/faker-ruby/faker](https://github.com/faker-ruby/faker)).
+
+Take a look into that search again… Don’t you see something special? We looked for ‘_darkness_’ and although we don’t have any matching result, we can see that the ones containing ‘_dark_’ were displayed. This is because of the text analyzers. They know these two words are not the same but they do have similar semantics.
+
+Pretty easy, wasn’t it?
+
+What happens if we look for ‘_dakr_’ instead of ‘_dark_’?
+
+![No results found](images/with_typos.png)
+
+Oops, no results. Let’s fix this.
+
+**Step 7: searching with typos and highlighting the match in searches’ results.**
+
+Our search will now look like this:
+
+```ruby
+def self.highlight(query)
+ search(
+   query: {
+     multi_match: {
+       query: query,
+       fuzziness: 'AUTO',
+       fields: %w[body]
```

Comment:
Did you remove the other field in the process?

Reply:
Oops, was missing in my project, but I'm fixing it.